src/Shared/TaskLoggingHelper.cs

```diff
@@ -503,6 +503,28 @@ public void LogMessageFromResources(MessageImportance importance, string message
 #endif
         }
 
+        /// <summary>
+        /// Logs a file generated from the given data.
+        /// </summary>
+        /// <param name="filePath">The file path.</param>
+        /// <param name="content">The content of the file.</param>
+        public void LogIncludeGeneratedFile(string filePath, string content)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(filePath, nameof(filePath));
+            ErrorUtilities.VerifyThrowArgumentNull(content, nameof(content));
+
+            var e = new ExtendedBuildMessageEventArgs("RESPONSEGENERATEDFILE", "", null, null, MessageImportance.Low)
+            {
+                ExtendedMetadata = new Dictionary<string, string>
+                {
+                    { "FILEPATH", filePath },
+                    { "CONTENT", content },
+                }
+            };
```

Comment:
I'm a bit concerned about using extended messages for this--I think of them primarily as an extensibility mechanism for third parties, not for MSBuild itself. While its name is a bit weird, could we extend `ResponseFileUsedEventArgs` to support inline content? Or add a new event type?

Reply:
Big agree with @rainersigwald on this note - especially now that we can add new event times in a non-breaking way due to binlog forward compatibility we should feel free to create or extend event types almost to excess!

Reply:
I agree. My initial design was to use new event `ResponseGeneratedFileUsedEventArgs` similar to `ResponseFileUsedEventArgs` but the build was failing https://github.com/dotnet/msbuild/runs/22434443891
```
Usage of unsecure BinaryFormatter during serialization of custom event type 'ResponseGeneratedFileUsedEventArgs'. This will be deprecated soon. Please use Extended*EventArgs instead. More info: https://aka.ms/msbuild/eventargs
```
Is it possible to turn that off?

Reply:
No, that rule is fixed now--but you can fix it! You'll have to add de/serialization code, for example similar to https://github.com/dotnet/msbuild/pull/8146/files#diff-02de8b63c474f29b5f0b4cebfef35896934503bb16bb83d2411dda65600a42f0R510

Reply:
Done